qemu/tests/check_link_speed_duplex.py

```diff
@@ -0,0 +1,186 @@
+        param method: the method to check, one of 'traceview' and 'powershell'
+        """
+        if method == 'traceview':
+            speed, duplex = get_speed_duplex_traceview()
+        elif method == 'powershell':
+            speed, duplex = get_speed_duplex_powershell(session)
+        else:
+            test.error("Method %s not supported", method)
+        if speed != tar_speed or duplex != tar_duplex:
+            test.fail("The speed and duplex is incorrect in %s, "
+                      "with speed=%s, duplex=%s" % (method, speed, duplex))
+
+    def get_speed_duplex_linux(session):
+        """
+        Get the speed and duplex information on linux guests.
+
+        return: a tuple of (speed, duplex), which speed is measured by mbps,
+                and duplex is one of 'half' and 'full'.
+        """
+        error_context.context("Get speed & duplex info", logging.info)
+        mac = vm.get_mac_address(0)
+        ethname = utils_net.get_linux_ifname(session, mac)
+        check_speed_cmd = params["check_speed_cmd"] % ethname
+        status, output = session.cmd_status_output(check_speed_cmd)
+        if status:
+            test.fail("Failed to get speed info,"
+                      "status=%s, ouput=%s" % (status, output))
+        logging.info(output)
+        result = re.findall(r"(?:Speed:\s+(\d+)Mb/s)|(?:Duplex:\s+(\w+))",
+                            output)
+        if len(result) < 2:
+            test.error("Can't get speed or duplex info")
+        speed = int(result[0][0])
+        duplex = result[1][1].lower()
+        return (speed, duplex)
+
+    def run_test_linux(session, tar_speed, tar_duplex):
+        """
+        Start the test on linux guests. Check whether the speed and duplex
+        information of the guest's nic matches the input params.
+
+        param tar_speed: target speed expected
+        param tar_duplex: target duplex expected
+        """
+        speed, duplex = get_speed_duplex_linux(session)
+        if speed != tar_speed or duplex != tar_duplex:
+            test.fail("The speed and duplex is incorrect, "
+                      "with speed=%s, duplex=%s" % (speed, duplex))
+
+    def run_test(session, tar_speed, tar_duplex):
+        """
+        Start the test according to the guest's os.
+
+        param tar_speed: target speed expected
+        param tar_duplex: target duplex expected
+        """
+        if os_type == "windows":
+            run_test_windows(session, tar_speed, tar_duplex)
+        elif os_type == "linux":
+            run_test_linux(session, tar_speed, tar_duplex)
+
+    timeout = params.get("timeout", 360)
+    os_type = params.get("os_type")
+    default_tar_speed = int(params.get("default_tar_speed"))
+    modify_tar_speed = int(params.get("modify_tar_speed"))
+    tar_duplex = params.get("tar_duplex")
+    modify_speed_param = params.get("modify_speed_param")
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
```

Comment:
Not necessary now, `vm.wait_for_login` will check it.

Reply:
Modified. Thanks